docs/modules/clients/pages/java.adoc

```diff
@@ -1950,6 +1899,64 @@ You can also configure it within the Spring context, as shown below:
 </beans>
 ----
 
+== Handling Failures
+
+There are three main failure cases and configurations you can perform to achieve proper behavior.
+
+**Handling Client Connection Failure:**
+
+While the client is trying to connect initially to one of the members in the
+`ClientNetworkConfig.addressList`, all the members might be not available.
+Instead of giving up, throwing an exception and stopping the client,
+the client retries to connect as configured which is described in the
+<<configuring-client-connection-retry, Configuring Client Connection Retry section>>.
+
+The client executes each operation through the already established connection to the cluster.
+If this connection(s) disconnects or drops, the client tries to reconnect as configured.
+
+**Handling Unresponsive Connections**
+
+It is possible for a connection to be open, but unresponsive, due to several scenarios like JVM pauses or TCP half-open sockets. If this occurs, a client which already has an open connection to that member will *not* be able to tell that the member is unresponsive because the connection is still open and no errors are received from the member side. Client operations will stay in-flight, potentially indefinitely, in this scenario. To avoid this, Hazelcast provides <<java-client-failure-detectors, Client Failure Detectors>> to mark members which appear to be unresponsive or down. Once the member is marked as suspect by the client, operations will fail as expected.
+
+**Handling Retry-able Operation Failure:**
+
+While sending the requests to related members, operations can fail due to various reasons.
+Read-only operations are retried by default. If you want to enable retry for the other operations,
+you can set the `redoOperation` to `true`. See the <<enabling-redo-operation, Enabling Redo Operation section>>.
+
+You can set a timeout for retrying the operations sent to a member.
+This can be provided by using the property `hazelcast.client.invocation.timeout.seconds` in `ClientProperties`.
+The client retries an operation within this given period if it is a read-only operation or
+you enabled the `redoOperation` as stated in the above paragraph.
+This timeout value is relevant only when there is a failure of the following types:
+
+* Member throws an exception
+* Connection between the client and member is closed
+* Client's heartbeat requests are timed out
+
+Unresponsive connections, described in the previous section, are one case where it may not be possible

```

Comment:
`Client's heartbeat requests are timed out` are same as unresponsive connections/failure detectors. This sounds more confusing to me right now.

Reply:
May be we should change the third item to something like following
* <<java-client-failure-detectors, Client Failure Detectors>> detects a member is unresponsive instead of `Client's heartbeat requests are timed out` 

And remove the next paragraph. The new paragraphs may sound like something we `can't` do if reader is not careful enough. But the listing above is already mentions that when we can take action.

Reply:
I agree with changing the third bullet point and removing the lines at 1937 & 1938. I'll update that in a moment.

Reply:
Wording has been updated.